### DISCUSSION TITLE
Master thesis about OpenCilk

### DISCUSSION THREAD

**Discussion by Lozan20** [OpenCilk/opencilk-project#169] [Q&A]
Hi Experts!
I'm in my senior year in AGH, University of Science and Technology in Kraków, Poland. Together with my professor that teaches about parallel programming and performance engineering, we took interest in OpenCilk. I want to write my master's thesis about OpenCilk that in the future will be used as a "guide" to new students in my professor's classroom. My main goal is to help people with OpenCilk. Many of my colleagues didn't like parallel programming, especially because of old fashioned tools like openmp or mpi, so i want to give new students something fresh. The main part of the master's thesis will be implementing some basic problems known in parallel programming, and then, using all the knowledge that I gained from that to explain everything as much as possible + of course some performance tests etc. If you have any tips & tricks, materials or anything else, your help will be very important to me!

**Comment by behoppe** [comment]
Hi @Lozan20 and welcome to OpenCilk. I lead user outreach and documentation for the team. I would be happy to help you make use of our existing documentation, and I'm excited to see how your guide shapes up. If you're agreeable, we'd love for that to become a community-wide resource.

I'd also like to introduce you to @timkaler and @ailiop. I think Tim and Alex will be interested in how you're implementing basic problems in parallel programming. Tim has been developing SpeedCode, a prototype online programming platform for learning how to write fast code; and Alex has been working on sample problems to populate the platform. SpeedCode (which will be ready soon) allows users to develop fast multicore solutions to short programming problems. I think/hope your work could fit nicely with theirs. Tim and Alex, could you please chime in?

**Comment by behoppe** [comment]
Hi @Lozan20 @timkaler @ailiop et al, I'm curious what you think of the "common core" of parallel programming as presented in this "[hands-on exploration of OpenMP common core](https://www.nersc.gov/assets/omp-common-core-short-Feb2018.pdf)."  They present one example problem in each of the following areas. Looks solid IMO. I would love for OpenCilk to have exercises like these.
* Creating Threads
* Synchronization
* Parallel Loops
* Data environment
* Memory model
* Irregular Parallelism and tasks

**Comment by behoppe** [comment]
Hi @Lozan20 @timkaler @ailiop, as we start curating problems, we might organize them using earlier work on *patterns for parallel programming*. What do you think? Please let me know how I can help build our collection of problems and tutorials.

More info below.

The text [Patterns for Parallel Programming](https://www.amazon.com/Patterns-Parallel-Programming-Software-ebook/dp/B001UG3IQQ) presents four general "design spaces" within the "pattern language" of parallel programming: finding concurrency, algorithm structure, supporting structure, and implementing mechanism. (Tim Mattson wrote this text and the OpenMP tutorial I mentioned previously -- I am hearing very good things about him in the CS ed community.) Some of the patterns are
* Algorithm structure
  - Task parallelism
  - Divide and conquer
  - Geometric decomposition
  - Recursive data
  - Pipeline
  - Event-based coordination
* Supporting structures
  - SPMD
  - Master/worker
  - Loop parallelism
  - Fork/join
  - Shared data

Other notable resources on patterns for parallel programming include
* [Our Pattern Language](https://patterns.eecs.berkeley.edu/) by UC Berkeley
* [Structured Parallel Programming: Patterns for Efficient Computation](https://www.amazon.com/Structured-Parallel-Programming-Efficient-Computation/dp/0124159931) 
* [Patterns of Parallel Programming with .NET](https://download.microsoft.com/download/3/4/D/34D13993-2132-4E04-AE48-53D3150057BD/Patterns_of_Parallel_Programming_CSharp.pdf)

Also related
* [Programming models for parallel computing](https://books.google.com/books?id=L6kCCwAAQBAJ&printsec=frontcover#v=onepage&q&f=false)

**Comment by behoppe** [comment]
Related to this thread, there is an issue for the OpenCilk.org website about creating a page filled with resources about learning software performance engineering, which includes an ever-growing list of books etc.: https://github.com/OpenCilk/www.opencilk.org/issues/212
